### ISSUE TITLE
False positive E0618 (Expected function, found `{ Some(()) }` )

### ISSUE THREAD

**Issue by neonaot** [intellij-rust/intellij-rust#9126]
## Environment

* **IntelliJ Rust plugin version:** 0.4.175.4772-222
* **Rust toolchain version:** 1.64.0-nightly (d68e7ebc3 2022-07-20) x86_64-apple-darwin
* **IDE name and version:** CLion 2022.2 (CL-222.3345.126)
* **Operating system:** macOS 11.4
* **Macro expansion engine:** new
* **Name resolution engine:** new
* **Additional experimental features:** org.rust.cargo.features.settings.gutter, org.rust.cargo.evaluate.build.scripts, org.rust.macros.proc

Originally posted in https://youtrack.jetbrains.com/issue/CPP-29939 

## Problem description
<img width="576" alt="image" src="https://user-images.githubusercontent.com/49211026/181483107-af1b0e76-4bb1-4d6b-9c9b-8e0bd847660e.png">

## Steps to reproduce
```rust
fn main() {
    let validation_result = || -> Option<()> {
        Some(()) // Expected function, found `{ Some(()) }` [E0618] 
    }();

    dbg!(validation_result);
}

```

**Comment by justxuewei** [comment]
I am suffering from this problem either, is there any workaround to fix that?

**Comment by vlad20012** [comment]
It looks like a bug with operator precedence

**Comment by justxuewei** [comment]
> It looks like a bug with operator precedence

Ok, a possible workaround is to rollback to [0.4.174.4743-222](https://plugins.jetbrains.com/plugin/8182-rust/versions/stable/194425), it works well for me.
